feat: replaced pysnmplib with lextudio's pysnmp v7.1.8

gmaiyani-crest opened the pull request:
# Description

Adding support of python v3.13 and migrating from pysnmplib to lextudio's pysnmp v7.1.8

- Replaced **pysnmplib** with lextudio’s **pysmp v7.1.8**
- Replaced  **pysnmp-pyasn1** to lextudio's **pyasn1**
- Replaced  **pysnmp-pysmi** to lextudio's **pysmi**
- Included pysnmpcrypto (maintained by lextudio's) for stronger crypto algorithms with pysnmp v6+
- Made SC4SNMP compatible with Python 3.13
- Updated SNMP walk implementation from `bulkCmd` to custom `multi_bulk_walk_cmd`
- Introduced `deploymentUser` in values.yaml to make the sc4snmp entirely compatible with  lextudio's pysnmp.
- Added `DEPLOYMENT_USER` variable for Docker Compose deployment.
- Introduces **multi_bulk_walk_cmd** to walk multiple OID trees simultaneously with lexicographic mode control per varbind.

## Notes

Added custom **multi_bulk_walk_cmd** function for Multi-Varbind SNMP Walking.

Why?
- **bulk_walk_cmd** supports only a single varbind at a time and already includes lexicographic mode handling.
However, to achieve the same behavior for multiple varbinds (from different OID subtrees), each varbind must be walked separately in sequence.
- Current **bulk_cmd** of lextuido's pysnmp does not support **lexicographic**  option as a result we can not use bulk_cmd of lextudio's pysnmp for the same purpose. 
- The new custom **multi_bulk_walk_cmd** addresses this by allowing multiple varbinds to be walked in parallel within a single session, maintaining lexicographic consistency while significantly reducing network roundtrips and improving overall efficiency as compare to previous `bulkCmd`.
- Feature request PR for adding multi_bulk_walk_cmd at Lextudio — raised here: [[PR-LINK](https://github.com/lextudio/pysnmp/pull/216)]

Key Changes
- Added multi_bulk_walk_cmd function for parallel walking of multiple OID subtrees
- Implemented lexicographic mode control for each varbind
- Restored etingof's bulkCmd multi-varbind with lexicographic capability with modern async/await patterns
- Implemented precise maxRows handling that counts actual OIDs (not iterations) with automatic truncation
- Added per-varbind completion tracking for independent state management

## Type of change

Please delete options that are not relevant.

- [x] Dependency update
- [x] New feature
- [x] Refactor/improvement
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist

- [x] My commit message is [conventional](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [x] I have run pre-commit on all files before creating the PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

ikheifets-splunk committed:
feat: mTLS for Splunk 10 (#1197)

* feat: mTLS support for Splunk 10

* docs: fix typo

---------

Co-authored-by: Olga <86965961+omrozowicz-splunk@users.noreply.github.com>
Co-authored-by: ajasnosz <ajasnosz@splunk.com>

srv-rr-github-token committed:
chore(release): 1.13.0-beta.1

# [1.13.0-beta.1](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.12.3...v1.13.0-beta.1) (2025-06-10)

### Features

* mTLS for Splunk 10 ([#1197](https://github.com/splunk/splunk-connect-for-snmp/issues/1197)) ([e8cb2a4](https://github.com/splunk/splunk-connect-for-snmp/commit/e8cb2a4a08c655d6353d5389a7bb98ddf694d4af))

renovate[bot] committed:
chore(deps): update docker/login-action action to v3.4.0 (#1166)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

ajasnosz committed:
fix: python and dependency update (#1200)

srv-rr-github-token committed:
chore(release): 1.13.0-beta.2

# [1.13.0-beta.2](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.1...v1.13.0-beta.2) (2025-06-10)

### Bug Fixes

* python and dependency update ([#1200](https://github.com/splunk/splunk-connect-for-snmp/issues/1200)) ([dd39338](https://github.com/splunk/splunk-connect-for-snmp/commit/dd3933873223bdb271f53e28c11255a1e0548252))

omrozowicz-splunk committed:
chore: add missing sudo in integration test script (#1207)

omrozowicz-splunk committed:
chore: update microk8s (#1206)

* chore: update microk8s
* chore: add changelog

renovate[bot] committed:
chore(deps): update dependency urllib3 to v2 [security] (#1205)

* chore(deps): update dependency urllib3 to v2 [security]

* test: test [run-int-tests]

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: omrozowicz-splunk <omrozowicz@splunk.com>

ajasnosz committed:
feat: add discovery for docker (#1208)

* feat: add discovery for docker

* chore: update replica count

* chore: update replica count

cwadhwani-splunk committed:
feat: Added deployment for k8s (#1204)

* feat: Added deployment for k8s

* Changed the queue name from autodiscover to discovery

omrozowicz-splunk committed:
fix: replace - with _ for metrics (#1213)

* fix: add option to change "-" to an "_" to be compatible with metric schema in Splunk

* fix: doc fix and docker-compose change

* fix: add docs

* chore: pre-commit run

* doc: refine

* doc: address PR comments, update CHANGELOG

srv-rr-github-token committed:
chore(release): 1.13.0-beta.3

# [1.13.0-beta.3](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.2...v1.13.0-beta.3) (2025-07-09)

### Bug Fixes

* replace - with _ for metrics ([#1213](https://github.com/splunk/splunk-connect-for-snmp/issues/1213)) ([cd44b56](https://github.com/splunk/splunk-connect-for-snmp/commit/cd44b561c5b6f24df62de77b20cad7895eba74d4))

ajasnosz committed:
feat: add partial walk (#1215)

srv-rr-github-token committed:
chore(release): 1.13.0-beta.4

# [1.13.0-beta.4](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.3...v1.13.0-beta.4) (2025-07-22)

### Features

* add partial walk ([#1215](https://github.com/splunk/splunk-connect-for-snmp/issues/1215)) ([0e1f32f](https://github.com/splunk/splunk-connect-for-snmp/commit/0e1f32f7769783e9f02731599302f848230546fb))

omrozowicz-splunk committed:
test: parallelised integration tests (#1220)

* test: add one more worker

* chore: add parallelism to docker-compose tests

* chore: move some polling tests to worker

omrozowicz-splunk committed:
fix: problem with casting strings of scientific notation (#1219)

* fix: problems with interpreting scientific value as float

* fix: unnecessary import

* fix: run pre-commit

* fix: check regex only if value is an instance of str

* fix: wrong function

* test: delete unnecessary test case

* test: rename test file

* chore: run pre-commit

* chore: update CHANGELOG

srv-rr-github-token committed:
chore(release): 1.13.0-beta.5

# [1.13.0-beta.5](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.4...v1.13.0-beta.5) (2025-07-30)

### Bug Fixes

* problem with casting strings of scientific notation ([#1219](https://github.com/splunk/splunk-connect-for-snmp/issues/1219)) ([80818aa](https://github.com/splunk/splunk-connect-for-snmp/commit/80818aacedde6974dcb890464e7416c49c9ba86b))

ajasnosz committed:
fix: add ipv6 flag for worker poller (#1218)

srv-rr-github-token committed:
chore(release): 1.13.0-beta.6

# [1.13.0-beta.6](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.5...v1.13.0-beta.6) (2025-07-31)

### Bug Fixes

* add ipv6 flag for worker poller ([#1218](https://github.com/splunk/splunk-connect-for-snmp/issues/1218)) ([57d44ae](https://github.com/splunk/splunk-connect-for-snmp/commit/57d44aed2816aff9c38ed76a74a50167722f752d))

renovate[bot] committed:
chore(deps): update dependency mkdocs-material to v9.6.16 (#1211)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

cwadhwani-splunk committed:
feat: Feat/autodiscovery application code (#1217)

* feat: initial commit with application code

* Added logic for mounting the final output csv file

* feat: addressed review comments

* chore: removed the init container and handled the volume mounting without it

* feat: added support for scaling discovery worker pods

* chore: Updated port validation

* feat: Added support for snmp version 3 in docker and handled volume mounting

* fix: Corrected the discovery paths in environment variable and removed unnecessary variables

omrozowicz-splunk committed:
chore: fix license accept issue in integration tests (#1227)

cwadhwani-splunk committed:
feat: added configurable parameter for SNMP request's retry and timeo… (#1224)

* feat: added configurable parameter for SNMP request's retry and timeout. Handled tasktimeout configurable parameter for discovery

* fix: handled race condition when discovery tasks tries to update the output csv file concurrently

* chore: Executed make render command with the latest yaml changes

* chore: removed --min-rate as nmap dynamically detects a suitable rate at which packets should be sent

* chore: incorporated the pr review changes

omrozowicz-splunk committed:
fix: logging (#1223)

* fix: disable mongo logs, fix traps logging

* fix: run pre-commit, fix conflicts

* fix: run pre-commit

* fix: make mongo logging configurable

* fix: make mongo logging configurable

* fix: add docs

* fix: fix traps helm template 

* fix: add CHANGELOG

srv-rr-github-token committed:
chore(release): 1.13.0-beta.7

# [1.13.0-beta.7](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.6...v1.13.0-beta.7) (2025-08-13)

### Bug Fixes

* logging ([#1223](https://github.com/splunk/splunk-connect-for-snmp/issues/1223)) ([a98987c](https://github.com/splunk/splunk-connect-for-snmp/commit/a98987c3bd3600c85e8f48b4d9e48596610e110e))

omrozowicz-splunk committed:
fix: update dependencies (#1229)

srv-rr-github-token committed:
chore(release): 1.13.0-beta.8

# [1.13.0-beta.8](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.7...v1.13.0-beta.8) (2025-08-13)

### Bug Fixes

* update dependencies ([#1229](https://github.com/splunk/splunk-connect-for-snmp/issues/1229)) ([fcf8805](https://github.com/splunk/splunk-connect-for-snmp/commit/fcf8805300f112b5772970010a40563155b0151e))

cwadhwani-splunk committed:
docs: Added doc changes for discovery feature (#1228)

* docs: Added doc changes for discovery feature

* docs: spell checks

* chore: resolved pr review comments

cwadhwani-splunk committed:
chore: removed expires from the shared_task decorator

omrozowicz-splunk committed:
fix: traps add engine contextId (#1230)

* fix: add decode method security engine ID

* chore: add missing import

* fix: unit tests

* fix: run pre-commit

* fix: update docker env file, docs and changelog

* fix: update method description

---------

Co-authored-by: ajasnosz <ajasnosz@splunk.com>

srv-rr-github-token committed:
chore(release): 1.13.0-beta.9

# [1.13.0-beta.9](https://github.com/splunk/splunk-connect-for-snmp/compare/v1.13.0-beta.8...v1.13.0-beta.9) (2025-08-14)

### Bug Fixes

* traps add engine contextId ([#1230](https://github.com/splunk/splunk-connect-for-snmp/issues/1230)) ([3dff9cd](https://github.com/splunk/splunk-connect-for-snmp/commit/3dff9cdb1cb62ea99b31648ea425fc539f604eea))

cwadhwani-splunk committed:
chore: Added architecture diagram which includes discovery feature (#1231)

* chore: Added architecture diagram which includes discovery feature

* chore: updated background of Discovery

cwadhwani-splunk committed:
Merge branch 'develop' of https://github.com/splunk/splunk-connect-for-snmp into feat/autodiscovery

cwadhwani-splunk committed:
Merge branch 'feat/autodiscovery' of https://github.com/splunk/splunk-connect-for-snmp into feat/autodiscovery

cwadhwani-splunk committed:
chore: executed pre-commit command to fix black related fix

cwadhwani-splunk committed:
chore: fixed issues raised by hook id mypy

cwadhwani-splunk committed:
chore: updated poetry lock file

cwadhwani-splunk committed:
chore: added build-base to install pandas and numpy

cwadhwani-splunk committed:
Feat/autodiscovery with csv package (#1236)

* chore: used csv package instead of pandas

* chore: quick fix

* chore: pre-commit changes

* chore: resolved review comments

cwadhwani-splunk committed:
chore: black related fix

cwadhwani-splunk committed:
feat: added dashboards for tracking discovery tasks (#1237)

* feat: added dashboards for tracking discovery tasks

* chore: added documentation for the dashboard update

cwadhwani-splunk committed:
chore: trigger tests [run-int-tests]

gmaiyani-crest committed:
feat: Adding support of python v3.13 and migrating from pysnmplib to lextudio's pysnmp v7.1.0

- Replaced pysnmplib with lextudio's pysmp v7.1.0
- Moved from pysnmp-pyasn1 to lextudio's pyasn1
- Moved from pysnmp-pysmi to lextudio's pysmi
- Included pysnmpcrypto (maintained by lextudio's) for stronger crypto algorithms to used with pysnmp v6+
- Updated SNMP walk implementation from bulkCmd to bulkWalkCmd
- Added concurrency support for bulkWalkCmd-based SNMP requests
- Made SC4SNMP compatible with Python 3.13
- Updated SNMP walk implementation from bulkCmd to bulkWalkCmd
- Added concurrency support for bulkWalkCmd-based SNMP requests
- Introduced deploymentUser in values.yaml
- Added DEPLOYMENT_USER env variable for Docker Compose deployment [run-int-tests]

omrozowicz-splunk commented on charts/splunk-connect-for-snmp/templates/scheduler/deployment.yaml:
```diff
@@ -43,6 +43,14 @@ spec:
               "celery", "beat",
             ]          
           env:
+            - name: USER
```
is it okay that all 4 env variables have `.Values.deploymentUser`? If it really should be the same value, can we pass it once and then assign 4 times in the code?

ajasnosz replied:
Could we also name the variable a little better? It is hard to tell to what that User is referring

omrozowicz-splunk commented on splunk_connect_for_snmp/snmp/const.py:
```diff
@@ -16,25 +16,25 @@
-    "SHA384": config.usmHMAC256SHA384AuthProtocol,
-    "SHA512": config.usmHMAC384SHA512AuthProtocol,
-    "NONE": config.usmNoAuthProtocol,
+    "MD5": config.USM_AUTH_HMAC96_MD5,
```
I trust this mapping is okay 😂

gmaiyani-crest replied:
it’s definitely going to work this time 😂 I went through it line by line like 20+ times

omrozowicz-splunk commented on splunk_connect_for_snmp/snmp/tasks.py:
```diff
@@ -102,15 +88,23 @@ def walk(self, **kwargs):
+        SnmpActionError,
+    ),
+    throws=(
+        SnmpActionError,
```
Is this okay to have this duplicated?

omrozowicz-splunk replied:
I know this was here before, but looks like this is redundant

gmaiyani-crest replied:
I think we can go with MongoLocked instead of SnmpActionError, i dont know why it was duplicated before but it should be MongoLockLocked or we can just remove it.

omrozowicz-splunk reviewed (commented):
Looks good, please check my comments ☺️ Are we on more or less the same level of unit tests coverage as before?

ajasnosz commented on .github/workflows/ci-ui-tests.yaml:
```diff
@@ -75,6 +75,7 @@ jobs:
           sudo snap install microk8s --classic --channel=1.33/stable
           sudo apt-get install snmp -y
           sudo apt-get install python3-dev -y
+          sudo apt-get install snmp-mibs-downloader -y
```
Why would we require the mibs downloader? 
The mibs we are using are installed with mibserver, is it required from pysnmp to provide that?

gmaiyani-crest replied:
yes, correct we don't required it. I'll update it

ajasnosz commented on integration_tests/automatic_setup_compose.sh:
```diff
@@ -44,7 +44,9 @@ deploy_poetry() {
   poetry install
   poetry add --group dev splunk-sdk
   poetry add --group dev splunklib
-  poetry add --group dev pysnmplib
```
Would it be possible to have outside value in one place that would specify the pysnmp's used version instead of hardcoding it in every file?

gmaiyani-crest committed:
refactor: remove redundant env vars (LOGNAME, LNAME, USERNAME) [run-int-tests]

Signed-off-by: gmaiyani-crest <gmaiyani@splunk.com>

gmaiyani-crest commented:
> Looks good, please check my comments ☺️ Are we on more or less the same level of unit tests coverage as before?

Yes, the coverage is the same as before.

ajasnosz commented on splunk_connect_for_snmp/snmp/exceptions.py:
```diff
@@ -2,3 +2,31 @@ class SnmpActionError(Exception):
+    else:
+        summary = base_message
+
+    return summary + "\n  " + "\n  ".join(messages)
```
How would the message look like after parsing? Could we rename `i` and `e` to something more meaningful?

ajasnosz commented on splunk_connect_for_snmp/snmp/manager.py:
```diff
@@ -379,82 +414,169 @@ def do_work(
+        :param metrics: dictionary to store metrics collected from SNMP responses
+        :param transport: SNMP transport target
+        :param varbinds_bulk: set of SNMP varbinds to query
+        :param walk: boolean flag indicating if it is a walk operation
```
could we name it in a way that it would respond to boolean, and it would suggest it is a flag value? 
for example is_walk?

ajasnosz commented on test/snmp/test_process_snmp_data.py:
```diff
@@ -11,47 +14,70 @@ class TestProcessSnmpData(TestCase):
-        varbind_mock2_1.getMibSymbol.return_value = "UDP-MIB", "next_metric", 1
-        varbind_mock2_1.prettyPrint.return_value = "some text2"
-        varbind_mock2_1.getOid.return_value = "9.8.7.6"
+        v1 = Mock()
```
could we keep the previous name, to be more explicit?

gmaiyani-crest committed:
chore: update the pysnmp to v7.1.8 and refactor the code [run-int-tests]

Signed-off-by: gmaiyani-crest <gmaiyani@splunk.com>

gmaiyani-crest committed:
chore: implemented multi_bulk_walk_cmd to improve performance

gmaiyani-crest committed:
chore: take the develop pull

cwadhwani-splunk committed:
chore: removed active check feature from auto discovery

gmaiyani-crest committed:
test: added test cases for multi_bulk_walk_cmd and refactor the code [run-int-tests].

cwadhwani-splunk committed:
chore: updated the documentation accordingly

cwadhwani-splunk committed:
Merge branch 'develop' of github.com:splunk/splunk-connect-for-snmp into feat/autodiscovery

gmaiyani-crest committed:
chore: update the loading mibs logic

cwadhwani-splunk committed:
chore: updated hash in poetry.lock file

cwadhwani-splunk committed:
chore: Running make render[run-int-tests]

cwadhwani-splunk committed:
chore: updating poetry lock [run-int-tests]

gmaiyani-crest committed:
feat: migrate auto-discovery
replace threadpool executor with asyncio.Semaphore for concurrency control

Signed-off-by: gmaiyani-crest <gmaiyani@splunk.com>

cwadhwani-splunk committed:
chore: refactoring the discovery manager

cwadhwani-splunk committed:
chore: updating integeration tests docker env[run-int-tests]

cwadhwani-splunk committed:
chore: updating integeration tests docker env[run-int-tests]

gmaiyani-crest committed:
refactor: move shared fields and validators into BaseRecord

gmaiyani-crest committed:
chore: update the naming convention

gmaiyani-crest committed:
merge: merged branch 'feat/migrate-pysnmplib-to-lextudio-pysnmp-v7.1.0' of github.com:splunk/splunk-connect-for-snmp into feat/migrate-autodiscovery

gmaiyani-crest committed:
merge: pull of branch 'feat/autodiscovery' of github.com:splunk/splunk-connect-for-snmp into feat/migrate-autodiscovery

gmaiyani-crest committed:
merge: merged branch 'feat/autodiscovery' of github.com:splunk/splunk-connect-for-snmp into feat/migrate-autodiscovery

gmaiyani-crest committed:
chore: migrate auto-discovery feature to make compatible with lextuio's pysnmp and python 3.13

- Migrate auto-discovery feature to make compitable with lextuio's pysnmp and python 3.13
- Replace the thread-pool executor with semaphore asyncio for concurrency control

gmaiyani-crest committed:
merge: pull latest changes from the develop branch
